internal/wal/wal_aof.go

```diff
@@ -65,31 +170,212 @@ func (w *WALAOF) LogCommand(c *cmd.DiceDBCmd) {
+		return
 	}
 
-	if err := w.file.Sync(); err != nil {
-		slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+	if w.writeMode == "fsync" {
+		if w.isBuffered && w.bufWriter != nil {
+			if err := w.bufWriter.Flush(); err != nil {
+				slog.Warn("failed to flush WAL buffer", slog.Any("error", err.Error()))
+				return
+			}
+		}
+		if err := w.file.Sync(); err != nil {
+			slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+			return
+		}
+	} else {
+		w.syncMutex.Lock()
+		defer w.syncMutex.Unlock()
+		if time.Since(w.lastSyncTime) >= w.syncInterval {
+			if w.isBuffered && w.bufWriter != nil {
+				if err := w.bufWriter.Flush(); err != nil {
+					slog.Warn("failed to flush WAL buffer", slog.Any("error", err.Error()))
+					return
+				}
+			}
+			if err := w.file.Sync(); err != nil {
+				slog.Warn("failed to sync WAL", slog.Any("error", err.Error()))
+				return
+			}
+			w.lastSyncTime = time.Now()
+		}
 	}
 
 	slog.Debug("logged command in WAL", slog.Any("command", c.Repr()))
 }
 
-func (w *WALAOF) Close() error {
-	if w.file == nil {
+func (w *WALAOF) rotateLogIfNeeded() error {
+	fileInfo, err := w.file.Stat()
+	if err != nil {
+		return err
+	}
+
+	var bufferedSize int64 = 0
+	if w.isBuffered && w.bufWriter != nil {
+		bufferedSize = int64(w.bufWriter.Buffered())
+	}
+
+	if fileInfo.Size()+bufferedSize >= w.maxFileSize {
+		return w.rotateLog()
+	}
+	return nil
+}
+
+func (w *WALAOF) rotateLog() error {
+	if w.isBuffered && w.bufWriter != nil {
+		if err := w.bufWriter.Flush(); err != nil {
+			return err
+		}
+	}
+	if err := w.file.Sync(); err != nil {
+		return err
+	}
+	if err := w.file.Close(); err != nil {
+		return err
+	}
+
+	w.currentSegmentIndex++
+	if w.currentSegmentIndex >= w.maxSegments {
+		if err := w.deleteOldestSegment(); err != nil {
+			return err
+		}
+	}
+
+	segmentFileName := fmt.Sprintf("%s%d", segmentPrefix, w.currentSegmentIndex)
+	path := filepath.Join(w.logDir, segmentFileName)
+
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+
+	w.file = file
+	if w.isBuffered {
+		w.bufWriter = bufio.NewWriterSize(w.file, 4096)
+		w.writer = w.bufWriter
+	} else {
+		w.writer = w.file
+	}
+
+	if w.rotationMode == "time" && w.rotationTimer != nil {
+		w.rotationTimer.Reset(w.rotationInterval)
+	}
+
+	return nil
+}
+
+func (w *WALAOF) deleteOldestSegment() error {
+	files, err := filepath.Glob(filepath.Join(w.logDir, segmentPrefix+"*"))
+	if err != nil {
+		return err
+	}
+
+	if len(files) == 0 {
 		return nil
 	}
-	return w.file.Close()
+
+	oldestSegmentFilePath, err := w.findOldestSegmentFile(files)
+	if err != nil {
+		return err
+	}
+
+	if err := os.Remove(oldestSegmentFilePath); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *WALAOF) findOldestSegmentFile(files []string) (string, error) {
+	oldestSegmentFilePath := ""
+	oldestSegmentID := math.MaxInt64
+	for _, file := range files {
```

Comment:
Do we need to loop through all the files to find the oldest segment file?
Once the process is initiated, the oldest and the latest segment file can be stored in memory.

Reply:
yes for current implementation it was needed. I am going to revamp the entire thing based on https://excalidraw.com/#json=5i9YJvPvBYtihEbC4YA_A,fUc_OmTJ8_FyOdh3vd3C4A